src/peft/tuners/lora.py

```diff
@@ -1145,91 +1131,98 @@ def __init__(
+
+            # Freezing the pre-trained weight matrix
+            self.weight.requires_grad = False
+
+            init_lora_weights = kwargs.pop("init_lora_weights", True)
+            self.update_layer(adapter_name, r, lora_alpha, lora_dropout, init_lora_weights)
+            self.active_adapter = adapter_name
+
+        def forward(self, x: torch.Tensor) -> torch.Tensor:
+            # note: logic differs from default Linear because merging is not supported
+            result = super().forward(x)
+
+            if (
+                self.disable_adapters
+                or (self.active_adapter not in self.lora_A.keys())
+                or (self.r[self.active_adapter] == 0)
             ):
-                bnb.nn.Linear4bit.__init__(
-                    self,
-                    in_features,
-                    out_features,
-                    bias=kwargs.get("bias", True),
-                    compute_dtype=kwargs.get("compute_dtype", torch.float32),
-                    compress_statistics=kwargs.get("compress_statistics", True),
-                    quant_type=kwargs.get("quant_type", "nf4"),
-                )
-                LoraLayer.__init__(self, in_features=in_features, out_features=out_features)
-
-                # Freezing the pre-trained weight matrix
-                self.weight.requires_grad = False
-
-                init_lora_weights = kwargs.pop("init_lora_weights", True)
-                self.update_layer(adapter_name, r, lora_alpha, lora_dropout, init_lora_weights)
-                self.active_adapter = adapter_name
-
-            def forward(self, x: torch.Tensor):
-                result = super().forward(x)
-
-                if self.disable_adapters or self.active_adapter not in self.lora_A.keys():
-                    return result
-                elif self.r[self.active_adapter] > 0:
-                    result = result.clone()
-                    if not torch.is_autocast_enabled():
-                        expected_dtype = result.dtype
-                        x = x.to(self.lora_A[self.active_adapter].weight.dtype)
-                        output = (
-                            self.lora_B[self.active_adapter](
-                                self.lora_A[self.active_adapter](self.lora_dropout[self.active_adapter](x))
-                            ).to(expected_dtype)
-                            * self.scaling[self.active_adapter]
-                        )
-                    else:
-                        output = (
-                            self.lora_B[self.active_adapter](
-                                self.lora_A[self.active_adapter](self.lora_dropout[self.active_adapter](x))
-                            )
-                            * self.scaling[self.active_adapter]
-                        )
-                    result += output
                 return result
 
+            # FIXME: is it correct that we clone here but not for int8?
+            result = result.clone()
```

Comment:
I see now, indeed this looks quite strange, I don't think we should clone here, I also realized we don't have 4bit training tests, I can maybe add a test with 4bit training and test if this works as expected

Reply:
Made: https://github.com/huggingface/peft/pull/834